File: Mobiles/ios/iOS-MobileO11yDemo/Features/CarConnectionAction/Models/CarCommunicationError.swift
//
//  CarCommunicationError.swift
//  iOS-MobileO11yDemo
//
//  Created by Robert Magnusson on 04.06.24.
//

import Foundation

enum CarCommunicationError: Error, CaseIterable, CustomStringConvertible {
    case timeout, randomEcuError, mysticMagicError

    var description: String {
            switch self {
            case .timeout:
                return "Communication Timeout"
            case .randomEcuError:
                return "Random ECU Error"
            case .mysticMagicError:
                return "Mystic Magic Error"
            }
        }
}

File: Mobiles/flutter/lib/core/car/domain_layer/car_communication_error.dart
import 'package:flutter_mobile_o11y_demo/core/utils/random_element.dart';

abstract class CarCommunicationError extends Error {
  static CarCommunicationError randomError() {
    final errors = [
      TimeOutError(),
      RandomEcuError(),
      MysticMagicError(),
    ];
    return errors.getRandomElement();
  }
}

class TimeOutError extends CarCommunicationError {
  @override
  String toString() => 'Communication Timeout';
}

class RandomEcuError extends CarCommunicationError {
  @override
  String toString() => 'Random ECU Error';
}

class MysticMagicError extends CarCommunicationError {
  @override
  String toString() => 'Mystic Magic Error';
}

class NotConnectedError extends CarCommunicationError {
  @override
  String toString() => 'Not Connected Error';
}
